src/Native/Runtime/arm/PInvoke.S

```diff
@@ -3,39 +3,82 @@
 // See the LICENSE file in the project root for more information.
 
 #include <unixasmmacros.inc>
+#include <AsmOffsets.inc>
 
 .syntax unified
 .thumb
 
-#ifdef _DEBUG
-.rodata
-AssertMsg:                              .asciz "__FILE__:%s: %s is not implemented\n"
-FileName:                               .asciz "PInvoke.S"
-RhpPInvokeName:                         .asciz "RhpPInvoke"
-RhpPInvokeReturnName:                   .asciz "RhpPInvokeReturn"
+.macro INLINE_GET_TLS_VAR Var, trashReg
+	        ldr	         r0, =\Var(tlsgd)
+	        bl	         __tls_get_addr@PLT
+	        ldr	         \trashReg, =\Var(dtpoff)
+	        add	         r0, \trashReg
+.endmacro
 
-.text
-.macro GEN_ASSERT_FUNC func
-        GEN_ASSERT AssertMsg, FileName, \func
-.endm
-#endif
-
-// TODO: Implement Arm support
+.macro INLINE_GETTHREAD trashReg
+	        // Inlined version of call C_FUNC(RhpGetThread)
+	        INLINE_GET_TLS_VAR tls_CurrentThread, \trashReg
+.endmacro
 
 //
 // RhpPInvoke
 //
+// IN:  R0: address of pinvoke frame
+//
+// This helper assumes that its callsite is as good to start the stackwalk as the actual PInvoke callsite.
+// The codegenerator must treat the callsite of this helper as GC triggering and generate the GC info for it.
+// Also, the codegenerator must ensure that there are no live GC references in callee saved registers.
+//
 NESTED_ENTRY RhpPInvoke, _TEXT, NoHandler
-#ifdef _DEBUG
-        GEN_ASSERT_FUNC RhpPInvokeName
-#endif
+	        PROLOG_PUSH "{r4-r9,lr}"
+
+	        mov	        r5, r0
+	        mov	        r8, sp // r13
```

Comment:
You do not need to store sp into temporary register. It is same through the whole function.